File: leet_BTREE_0814_pruning___dfs.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {TreeNode}

def prune_tree(node)
    return node if node == nil
    node.left = prune_tree(node.left)
    node.right = prune_tree(node.right)
    return nil if node.left == nil && node.right == nil && node.val == 0
    return node
end

File: leet_BTREE_0814_pruning___dfs.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func pruneTree(node *TreeNode) *TreeNode {
    if node == nil {return node}
    node.Left = pruneTree(node.Left)
    node.Right = pruneTree(node.Right)
    if node.Left == nil && node.Right == nil && node.Val == 0 {
        return nil
    }
    return node
}
